### ISSUE TITLE
[Processor] '@graphql-eslint/graphql' was not found

### ISSUE THREAD

**Issue by therealgilles** [dotansimha/graphql-eslint#292]
I'm trying to configure the plugin with its processor following the config in the README.md file and I get the following error:

`Error: ESLint configuration of processor in '.eslintrc.js#overrides[0]' is invalid: '@graphql-eslint/graphql' was not found.`

Here is my config:
```JS
  overrides: [
    {
      files: ['*.tsx', '*.ts', '*.jsx', '*.js'],
      processor: '@graphql-eslint/graphql',
    },
    {
      files: ['*.graphql'],
      parser: '@graphql-eslint/eslint-plugin',
      plugins: ['@graphql-eslint'],
      rules: {},
    },
  ],
```

Any advice would be appreciated :)

**Comment by therealgilles** [comment]
Looks like the plugin is not getting loaded properly (and eslint does not report it):
```
error: Error: Failed to load plugin '@graphql-eslint' declared in '.eslintrc.js#overrides[1]': Cannot find module 'graphql/validation/rules/ExecutableDefinitionsRule'
```

**Comment by dotansimha** [comment]
> Looks like the plugin is not getting loaded properly (and eslint does not report it):
> 
> ```
> error: Error: Failed to load plugin '@graphql-eslint' declared in '.eslintrc.js#overrides[1]': Cannot find module 'graphql/validation/rules/ExecutableDefinitionsRule'
> ```

Do you have `graphql` installed in your project? and is it being resolved into the same `node_modules` where `@graphql-eslint/eslint-plugin` is located? 
Also, which version of `graphql` do you use?

**Comment by therealgilles** [comment]
I do but the ExecutableDefinitionsRule module does not exist. I see the following:

```
node_modules/graphql/validation/rules/ExecutableDefinitions.d.ts
node_modules/graphql/validation/rules/ExecutableDefinitions.js
node_modules/graphql/validation/rules/ExecutableDefinitions.js.flow
node_modules/graphql/validation/rules/ExecutableDefinitions.mjs
```

**Comment by therealgilles** [comment]
It's GraphQL v14.7.0.

**Comment by therealgilles** [comment]
It seems this plugin is only compatible with GraphQL v15.*.

**Comment by therealgilles** [comment]
Please add a check/warning and/or documentation. I'm using GatsbyJS and it does not support v15.* yet.

**Comment by dotansimha** [comment]
Thanks @therealgilles . I'm working on a fix, this was not intended.

**Comment by dotansimha** [comment]
I tried to fix that in https://github.com/dotansimha/graphql-eslint/pull/298 . can you please try the alpha version? `@graphql-eslint/eslint-plugin@0.8.1-alpha-7440aa7.0`

**Comment by therealgilles** [comment]
Thanks @dotansimha. I was able to install the alpha version but when I try to run eslint on a graphql file with a syntax error, it reports nothing. Am I misunderstanding how it's supposed to work?

**Comment by therealgilles** [comment]
After I disabled the apollographql/eslint-plugin-graphql plugin, it is reporting the error :)

Now I am seeing the following in VSCode:
```
[Error - 10:23:31 AM] Error: ENOTDIR: not a directory, stat '.../src/pages/index.js/0_document.graphql'
```

Same thing for all my .js/.ts files. Any idea what's happening here? I obviously do not have any 0_document.graphql file. Did I misconfigure something?

**Comment by therealgilles** [comment]
Also not clear about how the plugin integrates with graphql-config. For the apollographql/eslint-plugin-graphql plugin, I had to add the following to eslintrc.js:

```JS
    'graphql/template-strings': [
      `error`,
      {
        // using graphql.config.js
        env: 'literal',
        projectName: `graphqlProject`,
      },
      {
        // using graphql.config.js
        env: 'apollo',
        projectName: `graphqlProject`,
      },
    ],
```

Does this plugin not require any such config?

**Comment by therealgilles** [comment]
I'm getting the same thing using the cmd line:

```JS
Oops! Something went wrong! :(

ESLint: 7.19.0

Error: ENOTDIR: not a directory, stat '.../src/components/image.js/0_document.graphql'
Occurred while linting .../src/components/image.js/0_document.graphql:0
    at Object.statSync (node:fs:1128:3)
    at isTypeSync (.../node_modules/prettier/third-party.js:9748:46)
    at getDirectorySync (.../node_modules/prettier/third-party.js:9802:62)
    at ExplorerSync.searchSync (.../node_modules/prettier/third-party.js:9950:66)
    at _resolveConfig (.../node_modules/prettier/index.js:25593:50)
    at Function.resolveConfig.sync (.../node_modules/prettier/index.js:25622:42)
    at Program (.../node_modules/eslint-plugin-prettier/eslint-plugin-prettier.js:167:40)
    at .../node_modules/eslint/lib/linter/safe-emitter.js:45:58
    at Array.forEach (<anonymous>)
    at Object.emit (.../node_modules/eslint/lib/linter/safe-emitter.js:45:38)
```

**Comment by therealgilles** [comment]
Does this point to a conflict between this plugin and prettier? It looks like the plugin is adding fake files to the list of files to analyze when extracting the graphql queries from js/jsx/ts/tsx files. But then prettier comes in and checks the file type.

**Comment by therealgilles** [comment]
Looks like it's related to https://github.com/dotansimha/graphql-eslint/issues/88, but I'm not sure what I'm supposed to add to the config to fix it, if anything.

**Comment by flarnie-chegg** [comment]
fwiw - @dotansimha I was running into this same error and installing @graphql-eslint/eslint-plugin@0.8.1-alpha-7440aa7.0 fixed it. Thanks. :)

**Comment by therealgilles** [comment]
Seems like the workaround for the ENOTDIR error is to turn off the prettier/prettier rule for graphql files. Not exactly sure if it has any adverse effect.

```JS
 overrides: [
    {
      files: ['*.tsx', '*.ts', '*.jsx', '*.js'],
      processor: '@graphql-eslint/graphql',
    },
    {
      files: ['*.graphql'],
      parser: '@graphql-eslint/eslint-plugin',
      plugins: ['@graphql-eslint'],
      rules: {
        'prettier/prettier': 'off', // FIXME: need to turn off prettier for now due to an eslint issue
      },
    },
  ],
```

**Comment by therealgilles** [comment]
Misc question: If I have a field that's not defined in the schema such as this

```JS
query myPostQuery {
  allPost() {
    nodes {
      title
      moo // <-- incorrect field
    }
  }
}
```

am I supposed to get an error while running eslint? (because I'm not) or is there a rule to turn on?

**Comment by therealgilles** [comment]
@dotansimha: Thank you for making the plugin compatible with GraphQL v14.7.0, much appreciated :)

Could you enlighten me about my incorrect field question? is there a rule I have to enable for that? My thinking was that as it violates the schema, I should not have to specify any rule for it to be reported as an error.

The vscode graphql extension points it out as a problem but nothing comes up when I run eslint on the file. Therefore I am thinking the schema is loaded properly but I'm missing something, maybe in the config.

PS: It's never quite clear whether my configuration is correct or not. It would be great to be able to get a report telling me what schema(s) and rule(s) were found by the tool. Let me know if this information is already available in some way. Thank you!

**Comment by dotansimha** [comment]
> Misc question: If I have a field that's not defined in the schema such as this
> 
> ```graphql
> query myPostQuery {
>   allPost() {
>     nodes {
>       title
>       moo // <-- incorrect field
>     }
>   }
> }
> ```
> 
> am I supposed to get an error while running eslint? (because I'm not) or is there a rule to turn on?

So `moo` doesn't exist in the schema? You need to use a rule for that. 
You can either use `validate-schema` (which uses GraphQL as-is), or standalone rules, like this one: https://github.com/dotansimha/graphql-eslint/blob/master/docs/rules/fields-on-correct-type.md

**Comment by therealgilles** [comment]
I guess I understand I need rules to preferred syntax but it seems rules for incorrect syntax ought to be enabled by default. If not, it would be great to have examples in the README file showing rules to enable to highlight syntax errors.

Thoughts?

**Comment by dotansimha** [comment]
> Thank you for your response!
> 
> I guess I understand I need rules for preferred syntax but it seems rules for incorrect syntax ought to be enabled by default. If not, it would be great to have examples in the README file showing rules to enable to highlight syntax errors.
> 
> Thoughts?

You are totally right. I didn't enable these by default is because most IDEs have GraphQL extensions or GraphiQL extensions (or, graphql-language-server) that would conflict with this validation, leading to a conflict or redundant warnings. I think using these might give you a better experience.

**Comment by JustinTRoss** [comment]
I get the same error as defined in the issue description.

Versions:
Eslint: 7.32.0
GraphQL: 15.5.2
Node: 14.17.5
@graphql-eslint/eslint-plugin: 2.4.0

Possibly relevant is that I'm using Yarn berry and thus have no node_modules.

**Comment by B2o5T** [comment]
Hi @JustinTRoss can you create a reproduction repo?

**Comment by JustinTRoss** [comment]
@B2o5T https://github.com/JustinTRoss/graphql-eslint-issue-repro

**Comment by B2o5T** [comment]
@JustinTRoss thanks for the reproduction, it's so much helpful 👍, you can try this version `2.4.1-alpha-7542cbd.0` that is compatible with yarn berry.

**Comment by JustinTRoss** [comment]
Thanks @B2o5T. I'm surprised the fix was so simple 😅 .

**Comment by dotansimha** [comment]
Fixed in @graphql-eslint/eslint-plugin@2.5.0

**Comment by avaly** [comment]
This broke for me in v2.5.0. It worked in v2.4.0.

I'm using yarn v1 with `node_modules`, not berry.

GraphQL: 15.7.0
ESLint: 8.2.0
Node.js: 14.18.1

**Comment by Burnett2k** [comment]
Hi there, I'm getting the error from the original description after adding in an overrides section to an existing .eslintrs.js file. I'm using yarn and can see `graphql` as well as `@graphql-eslint` both installed in node_modules.

```
overrides: [
    { files: ["*.js"], processor: "@graphql-eslint/graphql" },
    {
      files: ["*.graphql"],
      parser: "@graphql-eslint/eslint-plugin",
      plugins: ["@graphql-eslint"],
      rules: {
        "@graphql-eslint/known-type-names": "error"
      }
    }
  ]
```

When I run yarn lint which previously worked, I get `Error: ESLint configuration of processor in '.eslintrc.js#overrides[0]' is invalid: '@graphql-eslint/graphql' was not found.`.

Have any tips for debugging this issue? 

I only need the unique-operation-names rule (don't need to look at schema), but am finding the setup a bit cumbersome for just checking operation names. My gql queries are in *.js files and use the `gql` tag.

If I remove the first override which checks for js files, it runs without error. I'm in a yarn workspaces directory, but I've installed everything as root. Maybe I need to install for each yarn workspace?

**Comment by mikew** [comment]
I'm getting this error in my editor (VS Code), while `eslint` on the command line works fine?

In my editor it errors with

```
[Info  - 11:09:20 AM] ESLint server is starting.
[Info  - 11:09:20 AM] ESLint server running in node v18.15.0
[Info  - 11:09:20 AM] ESLint server is running.
[Info  - 11:09:21 AM] ESLint library loaded from: [redacted]/node_modules/.pnpm/eslint@7.31.0/node_modules/eslint/lib/api.js
Uncaught exception received.
Error: ESLint configuration of processor in '.eslintrc.cjs » [redacted]/eslint-config/graphql#overrides[0]' is invalid: '@graphql-eslint/graphql' was not found.
```

meanwhile in a shell / CI, it works fine:

```
$ ./node_modules/.bin/eslint src/components/auth/AuthHandler.jsx

```

And if I make an intentional error, it's reported:

```
$ ./node_modules/.bin/eslint src/components/auth/AuthHandler.jsx

[redacted]/src/components/auth/AuthHandler.jsx
  21:8  error  Cannot query field "lolnotfound" on type "ExtendJwtResponse"  @graphql-eslint/fields-on-correct-type

✖ 1 problem (1 error, 0 warnings)
```

```
$ pnpm --version
8.14.0
$ node --version
v20.10.0
$ ./node_modules/.bin/eslint --version
v7.31.0
```
